plugin/advisor/catalog/state.go

```diff
@@ -257,219 +253,296 @@ func (d *databaseState) FindTable(find *TableFind) *Table {
+		comment:      copyStringPointer(col.comment),
 	}
-	return res
 }
 
-func (column *columnState) convertToCatalog() *Column {
-	return &Column{
-		Name:         column.name,
-		Position:     column.position,
-		Default:      column.defaultValue,
-		Nullable:     column.nullable,
-		Type:         column.columnType,
-		CharacterSet: column.characterSet,
-		Collation:    column.collation,
-		Comment:      column.comment,
+// Nullable returns nullable for the column.
+func (col *ColumnState) Nullable() bool {
+	return col.nullable != nil && *col.nullable
+}
+
+// Type returns type for the column.
+func (col *ColumnState) Type() string {
+	if col.columnType != nil {
+		return *col.columnType
 	}
+	return ""
 }
 
-type viewState struct {
+type columnStateMap map[string]*ColumnState
+
+func (m columnStateMap) copy() columnStateMap {
+	res := make(columnStateMap)
+	for k, v := range m {
+		res[k] = v.copy()
+	}
+	return res
+}
+
+// ViewState is the state for walk-through.
+type ViewState struct {
 	name       string
-	definition string
-	comment    string
+	definition *string
+	comment    *string
 }
-type viewStateMap map[string]*viewState
 
-func (m viewStateMap) convertToCatalog() []*View {
-	var res []*View
-	for _, view := range m {
-		res = append(res, view.convertToCatalog())
+func (view *ViewState) copy() *ViewState {
+	return &ViewState{
+		name:       view.name,
+		definition: copyStringPointer(view.definition),
+		comment:    copyStringPointer(view.comment),
 	}
-	return res
 }
 
-func (view *viewState) convertToCatalog() *View {
-	return &View{
-		Name:       view.name,
-		Definition: view.definition,
-		Comment:    view.comment,
+type viewStateMap map[string]*ViewState
+
+func (m viewStateMap) copy() viewStateMap {
+	res := make(viewStateMap)
+	for k, v := range m {
+		res[k] = v.copy()
 	}
+	return res
 }
 
-type extensionState struct {
+// ExtensionState is the state for walk-through.
+type ExtensionState struct {
 	name        string
-	version     string
-	description string
+	version     *string
+	description *string
+}
+
+func (extension *ExtensionState) copy() *ExtensionState {
+	return &ExtensionState{
+		name:        extension.name,
+		version:     copyStringPointer(extension.version),
+		description: copyStringPointer(extension.description),
+	}
 }
-type extensionStateMap map[string]*extensionState
 
-func (m extensionStateMap) convertToCatalog() []*Extension {
-	var res []*Extension
-	for _, extension := range m {
-		res = append(res, extension.convertToCatalog())
+type extensionStateMap map[string]*ExtensionState
+
+func (m extensionStateMap) copy() extensionStateMap {
+	res := make(extensionStateMap)
+	for k, v := range m {
+		res[k] = v.copy()
 	}
 	return res
 }
 
-func (extension *extensionState) convertToCatalog() *Extension {
-	return &Extension{
-		Name:        extension.name,
-		Version:     extension.version,
-		Description: extension.description,
+func copyStringPointer(p *string) *string {
```

Comment:
Is there any reason to make a copy of the value?

Reply:
I would like to use a deep copy to avoid unintended modifications.